qiskit_experiments/data_processing/data_processor.py

```diff
@@ -0,0 +1,118 @@
+specifies a set of node_inputs that it accepts and a __node_output__ that it
+provides. The __node_output__ of each DataAction must be contained in the node_inputs
+of the following DataAction in the DataProcessor's list. DataProcessor.format_data()
+usually takes in one entry from the data property of an ExperimentData object
+(i.e. a dict containing metadata and memory keys and possibly counts, like the
+Result.data property) and produces a new dict containing the formatted data. The data
+passed to DataProcessor.format_data() is passed to the first DataAction and the
+output is passed on in turn to each DataAction. DataProcessor.format_data() returns
+the data produced by the last DataAction.
+"""
+
+from typing import Any, Dict, List, Tuple, Union
+
+from qiskit_experiments.data_processing.base import DataAction
+from qiskit_experiments.data_processing.exceptions import DataProcessorError
+
+
+class DataProcessor:
+    """
+    Defines the actions done on the measured data to bring it in a form usable
+    by the calibration analysis classes.
+    """
+
+    def __init__(self, data_actions: List[DataAction] = None):
+        """Create a chain of data processing actions.
+
+        Args:
+            data_actions: A list of data processing actions to construct this data processor with.
+                If None is given an empty DataProcessor will be created.
+        """
+        self._nodes = []
+
+        if data_actions:
+            for node in data_actions:
+                self.append(node)
+
+        self._history = []
+
+    @property
+    def history(self) -> List[Tuple[str, Dict[str, Any]]]:
+        """
+        Returns:
+            The history of the data processor. The ith tuple in the history corresponds to the
+            output of the ith node. Each tuple corresponds to (node name, data dict).
+        """
+        return self._history
+
+    def append(self, node: DataAction):
+        """
+        Append new data action node to this data processor.
+
+        Args:
+            node: A DataAction that will process the data.
+
+        Raises:
+            DataProcessorError: if the output of the last node does not match the input required
+                by the node to be appended.
+        """
+        if len(self._nodes) == 0:
+            self._nodes.append(node)
+        else:
+            if self._nodes[-1].__node_output__ not in node.node_inputs:
+                raise DataProcessorError(
+                    f"Output of node {self._nodes[-1]} is not an acceptable " f"input to {node}."
+                )
+
+            self._nodes.append(node)
+
+    def output_key(self) -> Union[str, None]:
+        """Return the key to look for in the data output by the processor."""
+
+        if len(self._nodes) > 0:
+            return self._nodes[-1].__node_output__
+
+        return None
+
+    def format_data(self, data: Dict[str, Any], save_history: bool = False) -> Dict[str, Any]:
```

Comment:
Another point related to the `output_key` above. If we imagine a data processor as a complete data processing chain, I think it would be more useful to make the return type of the call be the actual value stored in `output_key`, rather than a dict.

This would make it more plugginable as a callable too. For example in #19 I added a `data_processor: Callable` kwarg to one of the functions that expects a callable of signature `f(Dict[str, any]) -> Tuple[float, float]`.

If you still want to be able to chain data processors together we could do this by adding method for composing two dataprocessors (basically adding all the nodes from one to the other).

Also to return the dict structure you could use the `return_history` and make it a bit more configurable to only return specified stages of history dicts (ie if you have 5 nodes, you could pick which node outputs to return using something list index like with `return_history=-1` for the last, `return_history=0` for the first, `return_history=True` or `return_history=[0, 1, 2, 3, 4]` for all etc.

Reply:
See here: bd230b5

Reply:
We also have the option to now specify the nodes in the history with 1bc7c83